File: #150_Evaluate_Reverse_Polish_Notation/Solution.py
class Solution:
	# @param {string[]} tokens
	# @return {integer}
	def evalRPN(self, tokens):
		if not tokens or len(tokens) == 2:
			return 0
		if len(tokens) == 1:
			try:
				return int(tokens[0])
			except Exception, e:
				return 0
		stack = []
		for t in tokens:
			if t == '+':
				stack.append(stack.pop() + stack.pop())
			elif t == '-':
				i = stack.pop()
				j = stack.pop()
				stack.append(j - i)
			elif t == '*':
				stack.append(stack.pop() * stack.pop())
			elif t == '/':
				i = stack.pop()
				j = stack.pop()
				re = abs(j) / abs(i)
				if i * j < 0:
					re = -re
				stack.append(re)
			else:
				stack.append(int(t))
		return stack[0]

sol = Solution()
print sol.evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"])

File: #150_Evaluate_Reverse_Polish_Notation/Solution.java
import java.util.Deque;
import java.util.LinkedList;

public class Solution {
    public int evalRPN(String[] tokens) {
        if (tokens.length == 0) return 0;
        if (tokens.length == 1) 
        	return isOperator(tokens[0]) ? 0 : Integer.parseInt(tokens[0]);
        int len = tokens.length;
        Deque<Integer> stk = new LinkedList<Integer>();
        int result = -1;
        for (int i = 0; i < len; i++) {
        	String token = tokens[i];
        	if (isOperator(token)) {
        		int a = stk.pop();
        		int b = stk.pop();
        		result = compute(b, a, token);
        		// System.out.println("a :" + a + " b :" + b + " result: " + result);
        		stk.push(result);
        	}
        	else {
        		stk.push(Integer.parseInt(tokens[i]));
        	}
        }
        return result;
    }

    private boolean isOperator(String token) {
    	return token.equals("+") || token.equals("-") || token.equals("*") || token.equals("/");
    }

    private int compute(int a, int b, String opt) {
    	if (opt.equals("+")) return a + b;
    	if (opt.equals("-")) return a - b;
    	if (opt.equals("*")) return a * b;
    	if (opt.equals("/")) return a / b; 
    	return -1;
    }

    public static void main(String[] args) {
    	// Solution sol = new Solution();
    	// String[] tokens = {"4", "13", "5", "/", "+"};
    	// System.out.println(sol.evalRPN(tokens));
        System.out.println(6 / -132); 
    }
}